src/electron.js

```diff
@@ -170,41 +172,52 @@ app.on('activate', function() {
   }
 });
 
-ipcMain.on('addProcessId', (event, processId) => {
-  processIds.push(processId);
-});
+function attachIpcMainListeners(ipcMainHandle, notifyCallback) {
+  // Notify is used for testing the ipc calls
+  const notify = (evt, data) => {
+    if (notifyCallback) notifyCallback(evt, data);
+  };
 
-ipcMain.on('removeProcessId', (event, processId) => {
-  processIds = processIds.filter(id => id !== processId);
-});
+  ipcMainHandle.on('addProcessId', (event, processId) => {
+    processIds.push(processId);
+    notify('addProcessId', processIds);
+  });
 
-ipcMain.on('killAllRunningProcesses', event => {
-  if (processIds.length) {
-    event.preventDefault();
-    killAllRunningProcesses();
-  }
-  app.quit();
-});
+  ipcMainHandle.on('removeProcessId', (event, processId) => {
+    processIds = processIds.filter(id => id !== processId);
+    notify('removeProcessId', processIds);
+  });
 
-ipcMain.on('triggerClose', (e, proceed) => {
-  if (!proceed) {
-    // user aborted
-    emitAppWillClose = true; // reset flag
-    return;
-  }
+  ipcMainHandle.on('killAllRunningProcesses', async event => {
+    if (processIds.length) {
+      await killAllRunningProcesses();
+    }
+    app.quit();
+    notify('killAllRunningProcesses', processIds);
+  });
 
-  mainWindow.close();
-});
+  ipcMainHandle.on('triggerClose', (e, proceed) => {
+    if (!proceed) {
+      // user aborted
+      emitAppWillClose = true; // reset flag
+      return;
+    }
 
-const killAllRunningProcesses = () => {
-  try {
-    processIds.forEach(processId => {
-      killProcessId(processId);
+    mainWindow.close();
+  });
+}
 
-      // Remove the parent or any children PIDs from the list of tracked
-      // IDs, since they're killed now.
-      processIds = processIds.filter(id => id !== processId);
-    });
+const killAllRunningProcesses = async () => {
+  try {
+    await Promise.all(
+      processIds.map(processId => {
```

Comment:
Curious why change from `forEach` to `map` here?

Reply:
This is a strange pattern if you haven't seen it before. `Promise.all(promiseArr)` waits for every `Promise` in `promiseArr` to resolve before resolving itself. It should make more sense if I write it out without using `.map` notation:

```
const mapCallback = processId => {
  processIds = processIds.filter(id => id !== processId);
  return killProcessId(processId);
};

// processIds = [1, 2, 3];
await Promise.all([
  mapCallback(1),
  mapCallback(2),
  mapCallback(3),
]);
```

`killProcessId` returns a `Promise`, and we need to get that promise into the array that `Promise.all()` is using. This essentially boils down to:

```
await Promise.all([
  killProcessId(1),
  killProcessId(2),
  killProcessId(3),
]);
```

Does that help?

Reply:
Nice, I knew it was due to needing a return value but for some reason I wasn't connecting the dots as to what exactly it was returning 😆 Clever.